avr-hal-generic/src/usart_spi.rs

```diff
@@ -0,0 +1,95 @@
+        register_suffix: $n:expr,
+        sclk: $sclkpin:ty,
+        mosi: $mosipin:ty,
+        miso: $misopin:ty,
+        cs: $cspin:ty,
+    ) => {
+        $crate::paste::paste! {
+            // This is quite a messy way to get the doc string working properly... but it works!
+            #[doc = concat!("**Clock:** `", stringify!($sclkpin), "`<br>**MOSI:** `", stringify!($mosipin), "`<br> **MISO:** `", stringify!($misopin), "`<br> **CS:** `", stringify!($cspin), "`")]
+            pub type [<Usart $n Spi>] = avr_hal_generic::usart_spi::UsartSpi<$HAL, $USART_SPI, $sclkpin, $mosipin, $misopin, $cspin>;
+
+            impl $crate::spi::SpiOps<$HAL, $sclkpin, $mosipin, $misopin, $cspin> for $USART_SPI {
+                fn raw_setup(&mut self, settings: &crate::spi::Settings) {
+                    use $crate::hal::spi;
+
+                    // Setup control registers
+                    // We start by setting the UBBRn to 0
+                    self.[<ubrr $n>].write(|w| unsafe {w.bits(0)});
+
+                    // We have to translate the character size register into the 2 bits which are the MSB/LSB and the phase
+                    // 5 Bit Char = MSB and 1st
+                    // 6 Bit Char = MSB and 2nd
+                    // 7 Bit Char = LSB and 1st
+                    // 8 Bit Char = LSB and 2nd
+                    self.[<ucsr $n c>].write(|w| {
+                        w.[<umsel $n>]().spi_master();
+
+                        match settings.data_order {
+                            crate::spi::DataOrder::MostSignificantFirst => match settings.mode.phase {
+                                spi::Phase::CaptureOnFirstTransition => w.[<ucsz $n>]().chr5(),
+                                spi::Phase::CaptureOnSecondTransition => w.[<ucsz $n>]().chr6(),
+                            },
+                            crate::spi::DataOrder::LeastSignificantFirst => match settings.mode.phase {
+                                spi::Phase::CaptureOnFirstTransition => w.[<ucsz $n>]().chr7(),
+                                spi::Phase::CaptureOnSecondTransition => w.[<ucsz $n>]().chr8(),
+                            },
+                        };
+
+                        match settings.mode.polarity {
+                            spi::Polarity::IdleLow => w.[<ucpol $n>]().clear_bit(),
+                            spi::Polarity::IdleHigh => w.[<ucpol $n>]().set_bit(),
+                        }
+                    });
+
+                    // Enable receiver and transmitter, and also the rec interrupt.
+                    self.[<ucsr $n b>].write(|w| w
+                        .[<txen $n>]().set_bit()
+                        .[<rxen $n>]().set_bit()
+                        .[<rxcie $n>]().set_bit()
+                    );
+
+                    // Set the baudrate of the UBRRn, idk what it should be set to, so for now, it'll be set to 0
+                    self.[<ubrr $n>].write(|w| unsafe{w.bits(0)});
```

Comment:
UBRR sets the SPI clock frequency, so it should look something like this? I tested and it works, but I'm not sure if this is the best way to write the register in Rust.

```suggestion
                    // Set the clock divider for SPI clock.
                    self.[<ubrr $n>].write(|w| {
                        match settings.clock {
                            crate::spi::SerialClockRate::OscfOver2 => w.bits(0),
                            crate::spi::SerialClockRate::OscfOver4 => w.bits(1),
                            crate::spi::SerialClockRate::OscfOver8 => w.bits(3),
                            crate::spi::SerialClockRate::OscfOver16 => w.bits(7),
                            crate::spi::SerialClockRate::OscfOver32 => w.bits(15),
                            crate::spi::SerialClockRate::OscfOver64 => w.bits(31),
                            crate::spi::SerialClockRate::OscfOver128 => w.bits(63),
                        }
                    });
```

Reply:
If you are certain on the numbers, you can make use of the commit 65b304ec974a ("generic: spi: Assign discriminant values to clock rate enum") that I just pushed:

```rust
self.[<ubrr $n].write(|w| {
    2u8.pow(settings.clock as u32) - 1
});
```

Reply:
Yes, the numbers are correct for UBRR. The numbers do not apply to SPI peripheral, but there it is `match`ed, so no issue.